shared/js/content-scripts/block-cookie.js

```diff
@@ -28,72 +28,109 @@
-                    const setDate = new Date(dateStr)
-                    const nextWeekDate = new Date()
-                    nextWeekDate.setTime(nextWeekDate.getTime() + nextWeek)
-                    if (setDate > nextWeekDate) {
-                        comps[i] = 'max-age=' + nextWeek
-                    }
+    /**
+     * Apply an expiry policy to cookies set via document.cookie.
+     * @param {string} secret Used to detect messages sent from the extension content-script
+     */
+    function applyCookieExpiryPolicy (secret) {
+        const cookieSetter = document.__lookupSetter__('cookie')
+        const cookieGetter = document.__lookupGetter__('cookie')
+        const lineTest = /(\()?(http[^)]+):[0-9]+:[0-9]+(\))?/
 
-                    break
-                } else if (comps[i].trim().startsWith('max-age')) {
-                    expiryFound = true
-                    const ageStr = comps[i].trim().replace('max-age=', '').trim()
-                    const setDate = Number(ageStr)
-                    if (setDate > nextWeek) {
-                        comps[i] = 'max-age=' + nextWeek
-                    }
-
-                    break
+        // Listen for a message from the content script which will configure the policy for this context
+        const loadPolicy = new Promise((resolve) => {
+            const messageListener = (event) => {
+                if (event && event.isTrusted && event.data && event.data.source === secret) {
+                    resolve(event.data)
+                    window.removeEventListener('message', messageListener)
                 }
             }
-
-            if (!expiryFound) {
-                comps.push('max-age=' + nextWeek)
+            window.addEventListener('message', messageListener)
+        })
+        console.log('inject cookie test', cookieSetter, secret)
+        document.__defineSetter__('cookie', (value) => {
+            // call the native document.cookie implementation. This will set the cookie immediately
+            // if the value is valid. We will override this set later if the policy dictates that
+            // the expiry should be changed.
+            cookieSetter.apply(document, [value])
+            try {
+                // determine the origins of the scripts in the stack
+                const scriptOrigins = (new Error().stack.split('\n')).reduce((origins, line) => {
+                    const res = line.match(lineTest)
+                    if (res && res[2]) {
+                        origins.add(new URL(res[2]).hostname)
+                    }
+                    return origins
+                }, new Set())
+                // wait for config before doing same-site tests
+                loadPolicy.then(({ tabRegisteredDomain, policy }) => {
+                    const sameSiteScript = [...scriptOrigins].every((host) => host === tabRegisteredDomain || host.endsWith(`.${tabRegisteredDomain}`))
```

Comment:
I wonder if this should be `.every` or if we should look only at the last item (one that actually does the setting) or the first item (one that invokes the action) 🤔 Do we know how those stacks look like on various sites? e.g. what happens when some CDN hosted libraries are involved? This might be worth discussing a bit more in tech design.